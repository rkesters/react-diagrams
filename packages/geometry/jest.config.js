const sharedConfig = require('../../jest.config.js')(__dirname);
module.exports = {
	...sharedConfig,
	testEnvironment: 'jsdom',

};
