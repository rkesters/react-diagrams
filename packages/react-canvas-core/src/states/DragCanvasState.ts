import { CanvasEngine } from '..';
import { AbstractDisplacementState, AbstractDisplacementStateEvent } from '../core-state/AbstractDisplacementState';
import { State } from '../core-state/State';

export interface DragCanvasStateOptions {
	/**
	 * If enabled, the canvas is available to drag
	 */
	allowDrag?: boolean;
}

export class DragCanvasState extends AbstractDisplacementState {
	// store this as we drag the canvas
	initialCanvasX: number ;
	initialCanvasY: number ;
	config: DragCanvasStateOptions;

	constructor(engine: CanvasEngine, options: DragCanvasStateOptions = {}) {
		super(engine,{
			name: 'drag-canvas'
		});
		this.config = {
			allowDrag: true,
			...options
		};
		this.initialCanvasX = 0;
		this.initialCanvasY= 0;
	}

	async activated(prev: DragCanvasState) {
		super.activated(prev);
		this.engine.getModel().clearSelection();
		await this.engine.repaintCanvas(true);

		// we can block layer rendering because we are only targeting the transforms
		for (let layer of this.engine.getModel().getLayers()) {
			layer.allowRepaint(false);
		}

		this.initialCanvasX = this.engine.getModel().getOffsetX() ?? 0;
		this.initialCanvasY = this.engine.getModel().getOffsetY() ?? 0;
	}

	deactivated(next: State) {
		super.deactivated(next);
		for (let layer of this.engine.getModel().getLayers()) {
			layer.allowRepaint(true);
		}
	}

	fireMouseMoved(event: AbstractDisplacementStateEvent) {
		if (this.config.allowDrag) {
			this.engine
				.getModel()
				.setOffset(this.initialCanvasX + event.displacementX, this.initialCanvasY + event.displacementY);
			this.engine.repaintCanvas();
		}
	}
}
