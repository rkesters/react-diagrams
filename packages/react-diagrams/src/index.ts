import {
	DefaultDiagramState,
	DiagramEngine,
	DiagramModel,
	LinkLayerFactory,
	NodeLayerFactory
} from '@projectstorm/react-diagrams-core';
import {
	DefaultLabelFactory,
	DefaultLinkFactory,
	DefaultNodeFactory,
	DefaultPortFactory
} from '@projectstorm/react-diagrams-defaults';
import { PathFindingLinkFactory } from '@projectstorm/react-diagrams-routing';
import { SelectionBoxLayerFactory, CanvasEngineOptions } from '@projectstorm/react-canvas-core';

export * from '@projectstorm/react-diagrams-core';
export * from '@projectstorm/react-diagrams-defaults';
export * from '@projectstorm/react-diagrams-routing';

/**
 * Construct an engine with the defaults installed
 */
export default (model: DiagramModel, options: CanvasEngineOptions = {}): DiagramEngine => {
	const engine = new DiagramEngine(model,options);

	// register model factories
	engine.getLayerFactories().registerFactory(new NodeLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new LinkLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new SelectionBoxLayerFactory(engine));

	engine.getLabelFactories().registerFactory(new DefaultLabelFactory(engine));
	engine.getNodeFactories().registerFactory(new DefaultNodeFactory(engine)); // i cant figure out why
	engine.getLinkFactories().registerFactory(new DefaultLinkFactory(engine));
	engine.getLinkFactories().registerFactory(new PathFindingLinkFactory(engine));
	engine.getPortFactories().registerFactory(new DefaultPortFactory(engine));

	// register the default interaction behaviours
	engine.getStateMachine().pushState(new DefaultDiagramState(engine));
	return engine;
};
