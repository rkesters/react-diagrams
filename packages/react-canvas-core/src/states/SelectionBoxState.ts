import { MouseEvent, TouchEvent } from 'react';
import { AbstractDisplacementState, AbstractDisplacementStateEvent } from '../core-state/AbstractDisplacementState';
import { State } from '../core-state/State';
import { SelectionLayerModel } from '../entities/selection/SelectionLayerModel';
import { Point, Rectangle } from '@projectstorm/geometry';
import { ModelGeometryInterface } from '../core/ModelGeometryInterface';
import { CanvasEngine } from '..';

export interface IClientRect {
	left: number;
	top: number;
	right: number;
	bottom: number;
	width: number;
	height: number;
}

function isaMouseEvent(value: unknown): value is MouseEvent {
	return value instanceof MouseEvent
}

export class SelectionBoxState extends AbstractDisplacementState {
	layer: SelectionLayerModel ;

	constructor(engine: CanvasEngine) {
		super(engine, {
			name: 'selection-box'
		});
		this.layer = new SelectionLayerModel();
	}

	activated(previous: State) {
		super.activated(previous);
		this.engine.getModel().addLayer(this.layer);
	}

	deactivated(next: State) {
		super.deactivated(next);
		this.layer.remove();
		this.engine.repaintCanvas();
		this.layer = new SelectionLayerModel();
	}

	getBoxDimensions(stateEvent: AbstractDisplacementStateEvent): IClientRect {
		const event = stateEvent.event;

		const rel: Point = isaMouseEvent(event)  ? this.engine.getRelativePoint(event.clientX, event.clientY) :
		this.engine.getRelativePoint(event.touches[0].clientX, event.touches[0].clientY);


		return {
			left: rel.x > this.initialXRelative ? this.initialXRelative : rel.x,
			top: rel.y > this.initialYRelative ? this.initialYRelative : rel.y,
			width: Math.abs(rel.x - this.initialXRelative),
			height: Math.abs(rel.y - this.initialYRelative),
			right: rel.x < this.initialXRelative ? this.initialXRelative : rel.x,
			bottom: rel.y < this.initialYRelative ? this.initialYRelative : rel.y
		};
	}

	fireMouseMoved(event: AbstractDisplacementStateEvent) {
		this.layer.setBox(this.getBoxDimensions(event));

		const relative = this.engine.getRelativeMousePoint({
			clientX: this.initialX,
			clientY: this.initialY
		});
		if (event.virtualDisplacementX < 0) {
			relative.x -= Math.abs(event.virtualDisplacementX);
		}
		if (event.virtualDisplacementY < 0) {
			relative.y -= Math.abs(event.virtualDisplacementY);
		}
		const rect = new Rectangle(relative, Math.abs(event.virtualDisplacementX), Math.abs(event.virtualDisplacementY));

		for (let model of this.engine.getModel().getSelectionEntities()) {
			if ((model as unknown as ModelGeometryInterface).getBoundingBox) {
				const bounds = (model as unknown as ModelGeometryInterface).getBoundingBox();
				if (rect.containsPoint(bounds.getTopLeft()) && rect.containsPoint(bounds.getBottomRight())) {
					model.setSelected(true);
				} else {
					model.setSelected(false);
				}
			}
		}

		this.engine.repaintCanvas();
	}
}
