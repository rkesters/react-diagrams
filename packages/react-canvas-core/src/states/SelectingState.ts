import { State } from '../core-state/State';
import { Action, ActionEvent, InputType } from '../core-actions/Action';
import { SelectionBoxState } from './SelectionBoxState';
import { MouseEvent } from 'react';
import { CanvasEngine } from '..';

export class SelectingState extends State {
	constructor(engine: CanvasEngine) {
		super(engine, {
			name: 'selecting'
		});
		this.keys = ['shift'];

		this.registerAction(
			new Action({
				type: InputType.MOUSE_DOWN,
				fire: (event: ActionEvent<MouseEvent>) => {
					const element = this.engine.getActionEventBus().getModelForEvent(event);

					// go into a selection box on the canvas state
					if (!element) {
						this.transitionWithEvent(new SelectionBoxState(this.engine), event);
					} else {
						element.setSelected(true);
						this.engine.repaintCanvas();
					}
				}
			}, engine)
		);
	}
}
