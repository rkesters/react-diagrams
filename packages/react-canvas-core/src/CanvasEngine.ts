import { debounce } from 'lodash';
import { CanvasModel } from './entities/canvas/CanvasModel';
import { FactoryBank } from './core/FactoryBank';
import { AbstractReactFactory } from './core/AbstractReactFactory';
import { LayerModel } from './entities/layer/LayerModel';
import { BaseListener, BaseObserver } from './core/BaseObserver';
import { MouseEvent } from 'react';
import { BaseModel } from './core-models/BaseModel';
import { Point } from '@projectstorm/geometry';
import { ActionEventBus } from './core-actions/ActionEventBus';
import { ZoomCanvasAction } from './actions/ZoomCanvasAction';
import { DeleteItemsAction } from './actions/DeleteItemsAction';
import { StateMachine } from './core-state/StateMachine';
export interface CanvasEngineListener extends BaseListener {
	canvasReady(): void;

	repaintCanvas(): void;

	rendered(): void;
}

/**
 * Defines the CanvasEngine options
 */
export interface CanvasEngineOptions {
	registerDefaultDeleteItemsAction?: boolean;
	registerDefaultZoomCanvasAction?: boolean;
	/**
	 * Defines the debounce wait time in milliseconds if > 0
	 */
	repaintDebounceMs?: number;
}

export class CanvasEngine<
	L extends CanvasEngineListener = CanvasEngineListener,
	M extends CanvasModel = CanvasModel
> extends BaseObserver<L> {
	protected model: M ;
	protected layerFactories: FactoryBank<AbstractReactFactory<LayerModel>>;
	protected canvas: HTMLDivElement | undefined;
	protected eventBus: ActionEventBus;
	protected stateMachine: StateMachine;
	protected options: CanvasEngineOptions;

	constructor(model: M, options: CanvasEngineOptions = {}) {
		super();
		this.model = model;
		this.eventBus = new ActionEventBus(this);
		this.stateMachine = new StateMachine(this);
		this.layerFactories = new FactoryBank();
		this.registerFactoryBank(this.layerFactories);

		/**
		 * Overrides the standard options with the possible given options
		 */
		this.options = {
			registerDefaultDeleteItemsAction: true,
			registerDefaultZoomCanvasAction: true,
			repaintDebounceMs: 0,
			...options
		};
		if (this.options.registerDefaultZoomCanvasAction === true) {
			this.eventBus.registerAction(new ZoomCanvasAction({}, this));
		}
		if (this.options.registerDefaultDeleteItemsAction === true) {
			this.eventBus.registerAction(new DeleteItemsAction({}, this));
		}
	}

	getStateMachine() {
		return this.stateMachine;
	}

	getRelativeMousePoint(event: { clientX: number; clientY: number }): Point {
		const point = this.getRelativePoint(event.clientX, event.clientY);
		return new Point(
			(point.x - (this.model?.getOffsetX() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0),
			(point.y - (this.model?.getOffsetY() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0)
		);
	}

	getRelativePoint(x:number, y: number): Point {
		const canvasRect = this.canvas?.getBoundingClientRect();
		if (!canvasRect) {
			throw new Error(`Can't get relative point because canvas has not been defined.`)
		}
		return new Point(x - canvasRect.left, y - canvasRect.top);
	}

	registerFactoryBank(factory: FactoryBank<any>) {
		factory.registerListener({
			factoryAdded: (event) => {
				event.factory.setDiagramEngine(this);
			},
			factoryRemoved: (event) => {
				event.factory.setDiagramEngine(null);
			}, eventWillFire: ():void => {} , eventDidFire: ():void => {}
		});
	}

	getActionEventBus() {
		return this.eventBus;
	}

	getLayerFactories() {
		return this.layerFactories;
	}

	getFactoryForLayer<F extends AbstractReactFactory<LayerModel>>(layer: LayerModel | string | undefined) {
		if (!layer) { return;}
		if (typeof layer === 'string') {
			return this.layerFactories.getFactory(layer);
		}
		return this.layerFactories.getFactory(layer.getType());
	}

	setModel(model: M) {
		this.model = model;
		if (this.canvas) {
			requestAnimationFrame(() => {
				this.repaintCanvas();
			});
		}
	}

	getModel(): M {
		return this.model;
	}

	repaintCanvas(promise: true): Promise<any>;
	repaintCanvas(): void;
	repaintCanvas(promise?: true): Promise<any> | void {
		const repaintDebounceMs  = this.options.repaintDebounceMs ?? 0;

		/**
		 * The actual repaint function
		 */
		const repaint = () => {
			this.iterateListeners((listener) => {
				if (listener.repaintCanvas) {
					listener.repaintCanvas();
				}
			});
		};

		// if the `repaintDebounceMs` option is > 0, then apply the debounce
		let repaintFn = repaint;

		if (repaintDebounceMs > 0) {
			repaintFn = debounce(repaint, repaintDebounceMs);
		}

		if (promise) {
			return new Promise<void>((resolve) => {
				const l = this.registerListener({
					rendered: () => {
						resolve();
						l.deregister();
					}
				} as L);
				repaintFn();
			});
		}

		repaintFn();
	}

	setCanvas(canvas?: HTMLDivElement | null) {
		if (this.canvas !== canvas) {
			this.canvas = canvas ?? undefined;
			if (canvas) {
				this.fireEvent(undefined, 'canvasReady');
			}
		}
	}

	getCanvas() {
		return this.canvas;
	}

	getMouseElement(event: MouseEvent): BaseModel<any> | null | undefined  {
		return null;
	}

	zoomToFit() {
		if (!this.canvas) {
			throw new Error(`Can not zoom to fit because canvas is not defined`)
		}
		const xFactor = this.canvas.clientWidth / this.canvas.scrollWidth;
		const yFactor = this.canvas.clientHeight / this.canvas.scrollHeight;
		const zoomFactor = xFactor < yFactor ? xFactor : yFactor;

		this.model?.setZoomLevel((this.model?.getZoomLevel() ?? 1) * zoomFactor);
		this.model?.setOffset(0, 0);
		this.repaintCanvas();
	}
}
