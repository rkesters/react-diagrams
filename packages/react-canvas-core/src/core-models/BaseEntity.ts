import * as _ from 'lodash';
import { CanvasEngine } from '../CanvasEngine';
import { BaseEvent, BaseListener, BaseObserver, ParameterFirst } from '../core/BaseObserver';
import { Toolkit } from '../Toolkit';
import { BaseModel } from './BaseModel';

export interface BaseEntityEvent<T extends BaseEntity = BaseEntity> extends BaseEvent {
	entity?: T;
}

export interface LockChangedEvent<T extends BaseEntity = BaseEntity> extends BaseEntityEvent<T>  {
	locked: boolean;
}
export interface BaseEntityListener extends BaseListener {
	lockChanged(event: LockChangedEvent ): void;
}

export type BaseEntityType = 'node' | 'link' | 'port' | 'point';

export interface BaseEntityOptions {
	id?: string;
	locked?: boolean;
}

export type BaseEntityGenerics = {
	LISTENER: BaseEntityListener;
	OPTIONS: BaseEntityOptions;
};

export interface DeserializeEvent<S = ReturnType<BaseEntity['serialize']>> {
	engine: CanvasEngine;
	data: S;
	registerModel(model: BaseModel): void;
	getModel<T extends BaseModel>(id: string | null): Promise<T | null>;
}

type ff<T extends BaseEntityGenerics = BaseEntityGenerics> = ParameterFirst<T['LISTENER']['lockChanged'], BaseEvent>
const FF: ff = {locked: false, entity: {} as BaseEntity} ;
export class BaseEntity<T extends BaseEntityGenerics = BaseEntityGenerics> extends BaseObserver<T['LISTENER']> {
	protected options: T['OPTIONS'];

	protected id: string;
	constructor(options: T['OPTIONS'] = {}) {
		super();
		this.id = options.id ?? Toolkit.UID();
		this.options = {
			id: this.id,
			...options
		};
	}

	getOptions(): T['OPTIONS'] {
		return this.options;
	}

	getID(): string {
		return this.id;
	}

	doClone(lookupTable: { [s: string]: any } = {}, clone: any): void {
		/*noop*/
	}

	clone(lookupTable: { [s: string]: any } = {}) {
		// try and use an existing clone first
		if (lookupTable[this.id]) {
			return lookupTable[this.id];
		}
		let clone = _.cloneDeep(this);
		clone.options = {
			...this.options,
			id: Toolkit.UID()
		};
		clone.clearListeners();
		lookupTable[this.id] = clone;

		this.doClone(lookupTable, clone);
		return clone;
	}

	clearListeners() {
		this.listeners = {};
	}

	deserialize(event: DeserializeEvent<ReturnType<BaseEntity['serialize']>>) {
		this.options.id = event.data.id;
		this.options.locked = event.data.locked;
	}

	serialize() {
		return {
			id: this.options.id,
			locked: this.options.locked
		};
	}

	// fireEvent<Event extends keyof Methods<L, B>,>( k: keyof T['LISTENER'], event: L) {
	// 	super.fireEvent(
	// 		{
	// 			entity: this,
	// 			...event
	// 		},
	// 		k
	// 	);
	// }

	public isLocked(): boolean {
		return this.options.locked ?? false;
	}

	public setLocked(locked: boolean = true) {
		this.options.locked = locked;

		this.fireEvent(
			{ locked: true },
			'lockChanged'
		);
	}
}
