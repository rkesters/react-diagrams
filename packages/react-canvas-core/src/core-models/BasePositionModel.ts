import { BaseModel, BaseModelGenerics, BaseModelListener, BaseModelOptions } from './BaseModel';
import { BaseEntityEvent, DeserializeEvent } from './BaseEntity';
import { Point, Rectangle } from '@projectstorm/geometry';
import { ModelGeometryInterface } from '../core/ModelGeometryInterface';

export interface BasePositionModelListener extends BaseModelListener {
	positionChanged(event: BaseEntityEvent<BasePositionModel>): void;
}

export interface BasePositionModelOptions extends BaseModelOptions {
	position?: Point;
}

export interface BasePositionModelGenerics extends BaseModelGenerics {
	LISTENER: BasePositionModelListener;
	OPTIONS: BasePositionModelOptions;
}

export class BasePositionModel<G extends BasePositionModelGenerics = BasePositionModelGenerics>
	extends BaseModel<G>
	implements ModelGeometryInterface
{
	protected position: Point;

	constructor(options: G['OPTIONS']) {
		super(options);
		this.position = options.position || new Point(0, 0);
	}

	setPosition(point: Point): void;
	setPosition(x: number, y: number): void;
	setPosition(x: Point | number , y?: number | undefined ) {
		this.position = Point.isa(x) ? x : new Point(x,y as number);
		this.fireEvent({}, 'positionChanged');
	}

	getBoundingBox(): Rectangle {
		return new Rectangle(this.position, 0, 0);
	}

	deserialize(event: DeserializeEvent<ReturnType<BasePositionModel['serialize']>>) {
		super.deserialize(event);
		this.position = new Point(event.data.x, event.data.y);
	}

	serialize() {
		return {
			...super.serialize(),
			x: this.position.x,
			y: this.position.y
		};
	}

	getPosition(): Point {
		return this.position;
	}

	getX(): number {
		return this.position.x;
	}

	getY() {
		return this.position.y;
	}
}
