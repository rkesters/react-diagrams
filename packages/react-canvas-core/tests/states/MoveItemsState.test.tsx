import { SyntheticEvent } from 'react';
import { mocked } from 'ts-jest/utils';
import { CanvasEngine } from '../../src/CanvasEngine';
import { CanvasModel } from '../../src/entities/canvas/CanvasModel';
import { DragCanvasState } from '../../src/states/DragCanvasState';
import { MoveItemsState } from '../../src/states/MoveItemsState';
import { TestLayer } from '../support/impls/TestLayer';
import React from 'react';
import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BaseModel } from '../../src/core-models/BaseModel';
import { after } from 'lodash';
import { BasePositionModel } from '../../src/core-models/BasePositionModel';
import { Point } from '@projectstorm/geometry';

describe('MoveItemsState', () => {
	let engine = new CanvasEngine();
	let eventModel = new BaseModel({});
	let mouseDownEvent: any = { event: { type: 'mousedown' } as SyntheticEvent };
	beforeEach(() => {
		engine = new CanvasEngine();
		engine.setModel(new CanvasModel());
		render(<div className="canvas"></div>);
		const canvas: HTMLDivElement = document.querySelector('.canvas');
		engine.setCanvas(canvas);

		eventModel = new BaseModel({});
		eventModel.isSelected = jest.fn().mockImplementation(() => false);
		engine.getModel().clearSelection = jest.fn();
		engine.repaintCanvas = jest.fn();
		mouseDownEvent = { ...mouseDownEvent, model: eventModel } as any;
	});

	afterEach(cleanup);

	test('Constructor', () => {
		const state = new MoveItemsState();

		expect(state.getOptions()).toMatchSnapshot();
	});

	test('Action mousedown without selection', () => {
		const state = new MoveItemsState();
		state.setEngine(engine);

		state.activated({} as any);
		const mockedIsSelected = mocked(eventModel.isSelected);
		const mockedClearSelection = mocked(engine.getModel().clearSelection);
		const mockedRepaintCanvas = mocked(engine.repaintCanvas);

		engine.getActionEventBus().fireAction(mouseDownEvent);
		expect(mockedIsSelected).toHaveBeenCalled();
		expect(mockedClearSelection).toHaveBeenCalled();
		expect(mockedRepaintCanvas).toHaveBeenCalled();
		expect(eventModel.getOptions().selected).toEqual(true);
	});

	test('Action mousedown with selection', () => {
		const state = new MoveItemsState();
		state.setEngine(engine);

		state.activated({} as any);
		const mockedIsSelected = mocked(eventModel.isSelected);
		const mockedClearSelection = mocked(engine.getModel().clearSelection);
		const mockedRepaintCanvas = mocked(engine.repaintCanvas);
		mockedIsSelected.mockImplementation(() => true);

		engine.getActionEventBus().fireAction(mouseDownEvent);
		expect(mockedIsSelected).toHaveBeenCalled();
		expect(mockedClearSelection).not.toHaveBeenCalled();
		expect(mockedRepaintCanvas).toHaveBeenCalled();
		expect(eventModel.getOptions().selected).toEqual(true);
	});

	test('Action mousedown without model', () => {
		const state = new MoveItemsState();
		state.setEngine(engine);

		state.activated({} as any);
		const mockedIsSelected = mocked(eventModel.isSelected);
		const mockedClearSelection = mocked(engine.getModel().clearSelection);
		const mockedRepaintCanvas = mocked(engine.repaintCanvas);

		engine.getActionEventBus().fireAction({ ...mouseDownEvent, model: undefined });
		expect(mockedIsSelected).not.toHaveBeenCalled();
		expect(mockedClearSelection).not.toHaveBeenCalled();
		expect(mockedRepaintCanvas).not.toHaveBeenCalled();
		expect(eventModel.getOptions().selected).not.toEqual(true);
	});

	test('Fire Mouse Event', async () => {
		const state = new MoveItemsState();
		state.setEngine(engine);
		engine.getModel().setOffset(10, 20);
		const layer = new TestLayer();
		engine.getModel().addLayer(layer);

		const entities: [BasePositionModel, BasePositionModel, BasePositionModel, BaseModel] = [
			new BasePositionModel({}),
			new BasePositionModel({ locked: true }),
			new BasePositionModel({ locked: false }),
			new BaseModel({})
		];
		engine.getModel().getSelectedEntities = jest.fn().mockImplementation(() => entities);

		state.initialPositions = { [entities[0].getID()]: { point: new Point(0, 0), item: entities[0] } };
		state.fireMouseMoved({ virtualDisplacementX: 30, virtualDisplacementY: 50 } as any);

		expect(entities[0].getPosition()).toMatchInlineSnapshot(`
		Point {
		  "x": 30,
		  "y": 50,
		}
	`);
		expect(Object.keys(state.initialPositions)).toEqual([entities[0].getID(), entities[2].getID()]);
	});
});
