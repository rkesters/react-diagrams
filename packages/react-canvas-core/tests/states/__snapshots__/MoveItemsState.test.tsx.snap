// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoveItemsState Constructor 1`] = `
Object {
  "name": "move-items",
}
`;
